### PR TITLE
MDS-2847 bond audit history

### PR DESCRIPTION
**Pull request by JustinMacaulay**
Adds a table and functions to store a bond record every time information on it is edited.

### CODE REVIEW

File: services/core-api/app/api/securities/resources/bond.py [hunk 1]
```diff
@@ -89,6 +89,7 @@ def get(self, bond_guid):
     def put(self, bond_guid):
         #remove the amount from the request if it exists as it should not be editable.
         temp_bond = Bond.find_by_bond_guid(bond_guid)
+        temp_bond.save_bond_history()
```
Comment (exzizt) [inline services/core-api/app/api/securities/resources/bond.py:92]: What if the below operations fail? We'll have incorrectly saved bond history for a failed edit operation. I think this should happen at the end after the bond has been successfully modified.

Comment (JustinMacaulay) [reply services/core-api/app/api/securities/resources/bond.py:92]: If I want to save the history the way that I am with a class method the bond history save has to happen before the bond is modified

Comment (exzizt) [reply services/core-api/app/api/securities/resources/bond.py:92]: I guess an option is to then delete this `temp_bond` object in the `catch` block.

Comment (JustinMacaulay) [reply services/core-api/app/api/securities/resources/bond.py:92]: Temp_bond is just the bond before we modify it. So I dont want to delete it, the function lives on the bond model so it takes self and marshals it then loads it into the history table.

Comment (exzizt) [reply services/core-api/app/api/securities/resources/bond.py:92]: I understand. Do we want to create a bond history record if the PUT operation fails?

File: services/core-api/app/api/securities/resources/bond.py [hunk 2]
```diff
@@ -98,8 +99,11 @@ def put(self, bond_guid):
```
Comment (exzizt) [inline services/core-api/app/api/securities/resources/bond.py:99]: The catch of this block should delete the bond history object too

File: services/core-api/app/api/securities/resources/bond.py (hunk 1 above)
Comment (JustinMacaulay) [reply services/core-api/app/api/securities/resources/bond.py]: I have changed this and it should be good now if you can take another look

File: services/core-api/app/api/securities/resources/bond.py (hunk 2 above)
Comment (JustinMacaulay) [reply services/core-api/app/api/securities/resources/bond.py:99]: it does

File: services/core-api/app/api/securities/models/bond.py [hunk 3]
```diff
@@ -44,6 +48,17 @@ class _ModelSchema(Base._ModelSchema):
     def __repr__(self):
         return '<Bond %r>' % self.bond_guid
 
+    def save_bond_history(self):
+        bond_json = marshal(self, BOND)
```
Comment (Jsyro) [inline services/core-api/app/api/securities/models/bond.py:52]: can this use BondHistory._schema().dumps, this is currently using two different librarys for dumping and loading. Which is how the api works, but can you see if that's an easy change?

File: services/core-api/app/api/securities/resources/bond.py (hunk 2 above)
Comment (exzizt) [reply services/core-api/app/api/securities/resources/bond.py:99]: image
We need to do `history.delete()` between lines 97 and 98 too.

Comment (exzizt) [reply services/core-api/app/api/securities/resources/bond.py:99]: I would also raise a specific error on 106, maybe InternalServerError with the exception message?

File: services/core-api/app/api/securities/models/bond.py (hunk 3 above)
Comment (JustinMacaulay) [reply services/core-api/app/api/securities/models/bond.py:52]: I will try that

File: services/core-api/app/api/securities/resources/bond.py (hunk 2 above)
Comment (JustinMacaulay) [reply services/core-api/app/api/securities/resources/bond.py:100]: The second exception block around the save already raises a SQLAlchemy exception to I am just catching it again so I can delete the history and then re-raising the initial exception.

File: services/core-api/app/api/securities/models/bond_history.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    __tablename__ = "bond_history"
+
+    bond_history_id = db.Column(db.Integer, primary_key=True, server_default=FetchedValue())
+    bond_id = db.Column(db.Integer, db.ForeignKey('bond.bond_id'), nullable=False)
+    amount = db.Column(db.Numeric(14, 2), nullable=False)
+    bond_type_code = db.Column(db.String)
+    issue_date = db.Column(db.DateTime)
+    permit_no = db.Column(db.String)
+    permit_guid = db.Column(UUID(as_uuid=True))
+    payer = db.Column(db.String)
```
Comment (Jsyro) [inline services/core-api/app/api/securities/models/bond_history.py:23]: can this be payer_name, payer on the bond model is used and is different.

**Pull request merged**